uplink/piecestore/download.go

```diff
@@ -100,17 +104,25 @@ func (client *Client) Download(ctx context.Context, limit *pb.OrderLimit, pieceP
 func (client *Download) Read(data []byte) (read int, err error) {
 	ctx := client.ctx
 	defer mon.Task()(&ctx, "node: "+client.peer.ID.String()[0:8])(&err)
+
+	if client.closed {
+		return 0, io.ErrClosedPipe
+	}
+
 	for client.read < client.downloadSize {
 		// read from buffer
 		n, err := client.unread.Read(data)
 		client.read += int64(n)
 		read += n
 
-		// if we have an error or are pending for an error, avoid further communication
-		// however we should still finish reading the unread data.
-		if err != nil || client.unread.Errored() {
+		// if we have an error return the error
+		if err != nil {
 			return read, err
 		}
+		// if we are pending for an error, avoid further requests, but try to finish what's in unread buffer.
+		if client.unread.Errored() {
+			return read, nil
```

Comment:
why don't you return an error in this case?

Reply:
Because `unread` still contains valid data, that can be read until it hits the error.

Reply:
Maybe I am just misunderstanding then. I thought if `client.unread.Errored()` is true, the error was already hit. I also thought that `Read` should return an error, even when the number of bytes read is larger than 0.

Reply:
```
Let's say you have this stream of data buffered, which ends in error

|------|----------| <- `unread` contains this
. . . . . . . . . . . . . . E
|------| <-- `read` contains this

Even though unread contains an error, the reading doesn't end in the error...
So we shouldn't return the error until we have read all the valid data.
```

Reply:
Ok, that makes sense. It looks like in some cases, errors are added to `client.unread` which are never returned or added to `client.closingError`. And this block seems to be the only place where `client.unread.Errored()` is used except for calls to `client.unread.IncludeError()`. So I am assuming it is okay if some unread errors are not propagated outwards?

Reply:
`unread.Read` returns the error when it reads it.

Reply:
!!
makes sense